Feature: library working on app closed

CremaLuca opened the pull request:
The `SMSReceivedBroadcastReceiver` doesn't have a reference to the app service to be awakened, it retrieves the name of the service in preferences/memory and creates a new instance of it via Intent then starts it.
The app service that extends `SMSReceivedServiceListener` is more or less like the old `ReceivedListener`, and by default, it is not bound to the app because we don't know if the app is open or not, it's up to the user.

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/SMSReceivedBroadcastReceiver.java:
```diff
@@ -6,15 +6,22 @@
 import android.telephony.SmsMessage;
+import android.util.Log;
+
+import it.lucacrema.preferences.PreferencesManager;
```
The package's name is `com.eis.preferences`

CremaLuca replied:
Thanks, I had to edit the package name because I didn't want to keep my name on it, but the compiler forgot to tell me that I had to change on the smslibrary files too

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java:
```diff
@@ -9,17 +9,24 @@
 
 import java.lang.ref.WeakReference;
 
+import it.lucacrema.preferences.PreferencesManager;
```
Same as above, the actual name is `com.eis.preferences`

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/SMSReceivedBroadcastReceiver.java:
```diff
@@ -60,4 +66,27 @@ private SmsMessage createMessageFromPdu(Object smsData, String format) {
+
+        } else {
+            //Listener null, nothing to perform
+        }
```
This `else` is useless, it should be removed

giovannivelludo commented on preferences/src/androidTest/java/com/eis/preferences/PreferencesManagerTest.java:
```diff
@@ -0,0 +1,162 @@
+        } catch (IOException e) {
+            Assert.fail("Should not have thrown an IOException");
+        }
+        Assert.assertEquals(DEFAULT_OBJECT_VALUE, PreferencesManager.<String>getObject(ctx, DEFAULT_OBJECT_KEY));
```
I ran this test on my machine and it failed at line 80:
```
java.lang.AssertionError: expected:<Test String> but was:<null>
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.failNotEquals(Assert.java:834)
at org.junit.Assert.assertEquals(Assert.java:118)
at org.junit.Assert.assertEquals(Assert.java:144)
at com.eis.preferences.PreferencesManagerTest.setObject_getBObject_isEquals(PreferencesManagerTest.java:80)
```

CremaLuca replied:
Forgot to clean the PreferencesManager class, the getObject and setObject were not ready to release yet. I removed the two methods and the testing methods

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/background/SMSKeepAliveService.java:
```diff
@@ -0,0 +1,52 @@
+     * Registers the broadcastReceiver on this service, so that it will also be called
+     * when the app is closed
+     */
+    private void registerMessageBroadcastReceiver() {
```
There's one thing I don't understand: we already have a registered SMSReceivedBroadcastReceiver, because it's declared in the manifest of smslibrary, and since it is a BroadcastReceiver it will always be called when a message is received, even if the app was destroyed. Then what's the point of having an SMSKeepAliveService that registers another BroadcastReceiver, when we already have one enabled?

CremaLuca replied:
Because our tests showed that BroadcastReceiver was not being called on most devices if the app was dead or if the phone was restarted, so as suggested in a few [StackOverflow answers ](https://stackoverflow.com/questions/16824341/keep-broadcast-receiver-running-after-application-is-closed) we decided to create the `SMSKeepAliveService`

giovannivelludo replied:
If you read the documentation I linked you'll see that there's a difference between BroadcastReceivers registered in the manifest and those registered with a Context:
in the first case, they are called even when the app is dead; in the second case, they're only called when their Context is active, and the Stackoverflow answer you posted refers to this second case. Our library's case is the first, so a service is not necessary.

CremaLuca replied:
What about our tests? We tested with a service to be started on message arrival that had to print a notification, and it wasn't working properly every time: when we closed the app the notification would appear only if the message arrived within a few minutes from when the app was closed.

RicDeZen replied:
Manifest declared Receivers are supposed to work no matter what happens to the app. During the demonstration of group's 1 demo for the killer app another activity was started in addition to the killer app, that was due to a Receiver activating from a dumb test app of mine, days upon days after installing it and after various cold boots of the emulator. A Receiver registered through a context will die out the moment the activity/application (depends on which you pass) is destroyed. I don't know what kind of tests you guys did but I think we need to delve further into this.

giovannivelludo replied:
I don't know what tests you did, and I'm sorry if I sound rude, but I trust Android's documentation more than you. This afternoon I'll try writing a simple app to receive messages using this library and I'll see what happens.

CremaLuca replied:
Ok, if the two of you are 100% sure about this I'll remove the service.

giovannivelludo replied:
I'm not 100% sure, I'd like to try the library and fix the BroadcastReceiver registered in the manifest first (if it doesn't work), if that is not possible then I'll accept the service. Even though it's not guaranteed that it will always run in the background since Android can kill it if it consumes too much battery, it's still better than SMS reception not working in the background at all.

giovannivelludo replied:
It seems that in order for the BroadcastReceiver to work it should be declared in the app's manifest, not in the library's manifest

RicDeZen replied:
This sounds very weird, I thought the manifests got merged while compiling

giovannivelludo replied:
@CremaLuca could you create another branch in your fork starting from Cogno-IDU:master and add there the application you were using to test the BroadcastReceiver?

giovannivelludo replied:
Here's a test app with a BroadcastReceiver that opens the main activity everytime a message is received, as you can see it works even after closing the app and after a reboot: [https://youtu.be/X4wUtfc9-vM](https://youtu.be/X4wUtfc9-vM)

CremaLuca replied:
@giovannivelludo I removed the KeepAliveService because it should be useless and needs more research, but proving your point on an emulator is not the way I'd do it. We tested on real physical phones and Huawei seems to have a few problems every now and then about the BroadcastReceiver.

I would keep it like this until this gets merged and until I have some more time to test, if this needs the service again I'll do another pull request to add it.

giovannivelludo replied:
Could you point me to the branch in your repo where you developed the app you were testing?

CremaLuca commented:
`Preferences` module has nothing to do with the `smslibrary` and I think it should be kept separate

giovannivelludo commented:
How does it have nothing to do with the library? They're both supposed to be used by apps that use our smslibrary, right? Otherwise there would be no reason to add it to this repository.
Here is an explanation of what Android modules are used for: [https://developer.android.com/studio/projects/](https://developer.android.com/studio/projects/)

CremaLuca commented:
>A module is a collection of source files and build settings that allow you to divide your project into discrete units of functionality

Saving configurations in memory is the same functionality as sending SMSs

giovannivelludo commented:
Then by your reasoning we might as well separate sending SMSes and receiving SMSes in two different modules, because they're two different functions.

CremaLuca commented:
@giovannivelludo as you have suggested I've removed the module and added the requested library files as a dependency from Jitpack. You were right, this is definitely a better choice than a whole module.

Cogno-IDU commented:
Awesome!

giovannivelludo commented:
Making `preferences` an external library only makes it more difficult to mantain, since now other groups can't modify it by making pull requests to this repo and you're the only one controlling it. Professor Peserico never said that the library we had to make for our app had to include only functionality relative to sending and receiving messages, so I don't get why you didn't just add class PreferencesManager to our library in the first place.

CremaLuca commented:
Actually `preferences` is stored in a public repository, so anyone can make pull requests to change it if needed (even though it's a very light class). I still don't agree that it should be placed inside the library module, it's reusable code and every edit you want to do on it should be propagated to every code using it (like `killer-app` is using this same library). What you could want to do is extend the class or make a wrapper around it so that only the requested methods are shown, but I don't think it's worth it right now.

CremaLuca commented:
Would it make sense to keep the normal listener in case the user just wants to use received messages only when the app is still open? Because as it is right now every code using the listener would be unusable. Do you think I should keep the possibility to keep the old listener?

giovannivelludo commented:
> Actually preferences is stored in a public repository, so anyone can make pull requests to change it if needed (even though it's a very light class).

You realize that if every time somebody wanted to add a class to this library they made another module and imported that instead, we would end up with too many different repositories to keep track of? Having preferences in another repository doesn't provide any advantage regarding pull requests, because if it was in this repository it could still be modified by opening pull requests.

> I still don't agree that it should be placed inside the library module, it's reusable code 

It would be reusable code even when placed inside this library. Right now the only app using PreferencesManager would be killer-app, and maybe network-dictionary if we merge that in the future, and they both would use this library.

> and every edit you want to do on it should be propagated to every code using it (like killer-app is using this same library).

That's exactly what would happen when placing PreferencesManager inside this library, you wouldn't even need to import it because killer-app already used this library. You're just adding reliance on an external service (jitpack).

> What you could want to do is extend the class or make a wrapper around it so that only the requested methods are shown, but I don't think it's worth it right now.

Me neither.  
I just don't understand your reasoning: when you created PreferencesManager, clearly the simplest thing you could do was simply putting it inside the library module of this project. Any extra steps should have a good reason to be taken, and so far nobody has come up with one. We already agreed that having a module for every set of classes providing different functionality doesn't make sense, because it provides no benefit compared to having those classes in the same package or different subpackages of the `smslibrary` module, it only adds bloat in the form of useless build.gradle and resource files, and it sets a precedent for people who might want to separate every bit of functionality of this library in different modules.

CremaLuca commented:
> You realize that if every time somebody wanted to add a class to this library they made another module and imported that instead

Yes and I think that's part of the process to give value to every step of the development, it's code that can be used again and again, and it's nowhere near dependent from this library in any way.

>It would be reusable code even when placed inside this library.

`import com.eis.smslibrary.PreferencesManager` ... What does the name `smslibrary` have to do with `preferences`? How can you ever think they can be anyhow related?

>Any extra steps should have a good reason to be taken, and so far nobody has come up with one

I did, I came up with two good ones, so...
1. Easily reusable code
2. Has nothing to do with sms library

giovannivelludo commented:
> Yes and I think that's part of the process to give value to every step of the development, it's code that can be used again and again, and it's nowhere near dependent from this library in any way.

SMSCore is also not dependent on this library in any way since it only uses system libraries and functions, do you think that reason is enough to move it to another module? As I already said, having PreferencesManager in this library wouldn't make a single difference in re-usability for our purposes, since killer-app already imports this library.

> `import com.eis.smslibrary.PreferencesManager` ... What does the name `smslibrary` have to do with `preferences`?

Nothing, however I don't see why semantic similarity between packages and classes' names should be one of the criteria used to decide whether certain classes should belong to certain packages. See for example class [`IsolatedContext`](https://developer.android.com/reference/android/test/IsolatedContext.html#public-methods) in package `android.test`, name `test` has nothing to do with `IsolatedContext`.
The criteria should be classes in the same package having related functionality, for example in the case of this library they should provide functions needed by our killer-app.

> How can you ever think they can be anyhow related?

They're related because they're both used by our killer app.

> Easily reusable code

You haven't explained how having a different module containing only PreferencesManager would make that class more reusable compared to just including it in this library.

> Has nothing to do with sms library

This is not true, PreferencesManager is used by classes in smslibrary, specifically SMSHandler and SMSReceivedBroadcastReceiver.

Cogno-IDU commented:
The PreferenceManager class is a class which has nothing to do with the sms library, can we agree on that? Not that it's used or not, that it has nothing to do with the library. Can we agree on this?

I'm gonna say it again more specifically, I'm not talking about if the class is used or where it's used, I'm talking about this:
We have a module which has one specific role: working with sms messages (sending them, receiving them, ...). Does PreferenceManager have anything to do with that?
I'm saying it doesn't, can we both agree on this @giovannivelludo ?

AlbertoUrsino commented on smslibrary/src/main/java/com/eis/smslibrary/listeners/SMSReceivedServiceListener.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Handles the service call, extracts the message from the intent extras and calls {@link #onMessageReceived(SMSMessage)}
+     *
+     * @param intent intent passed from the broadcastReceiver
```
Repetition

Cogno-IDU replied:
Honestly I don't think being so picky about param documentation "repetitions" is so necessary. I remember you that in the documentation it becomes `@param intent - intent passed from the broadcastReceiver` which in my opinion is better than `@param intent - passed from the broadcastReceiver`. I would even go as far as writing it as `@param intent - The intent passed from the broadcastReceiver`

AlbertoUrsino replied:
I did not know this fact of documentation.
Thanks to having pointed this out to me, now I feel even more picky.
However, I welcome your correction: `@param intent - The intent passed from the broadcastReceiver`

Cogno-IDU replied:
I was "picky" too, preferring the last solution, but since I found out (thanks to @CremaLuca ) that even in the official android api it would be sometimes written as `@param intent - passed from the broadcastReceiver` That I've just learned to accept all 3 solutions, (even though I'll still write the last one 😄 )

AlbertoUrsino commented on smslibrary/src/main/java/com/eis/smslibrary/SMSReceivedBroadcastReceiver.java:
```diff
@@ -60,4 +65,26 @@ private SmsMessage createMessageFromPdu(Object smsData, String format) {
+     * Calls the current subscribed app service
+     *
+     * @param context broadcast current context
+     * @param message received message
```
Repetition

AlbertoUrsino commented:
The code seems perfect except for some repetition between the specifications but it is because I am a bit fussy.

I followed the "problem" raised by @velludo very carefully and to give a small opinion about it I feel I must agree with @luca, not because it is in my group or other, but because "preferences" is so abstract from the library sms that makes no sense to put it in.

Apart from this, I will tell you one thing: I'm working on a project of mine that I won't tell you otherwise you stole it (a joke) in which I needed to save data in memory (access passwords) and for convenience I found myself "stealing" the "preferences" module from the @luca repository.

This is a small proof that it would be better to keep him busy with our library. Heck is useful as it is!

giovannivelludo commented:
> The PreferenceManager class is a class which has nothing to do with the sms library, can we agree on that? Not that it's used or not, that it has nothing to do with the library.

"Having nothing to do" [means that it has no relationship with the library](https://idioms.thefreedictionary.com/have+nothing+to+do+with). Clearly this is not true, as PreferencesManager is used and imported in the library, therefore they have a relationship, therefore they have something to do with each other.
When you say "Not that it's used or not, that it has nothing to do with the library" you're contradicting yourself, because if PreferencesManager is used by the library, that implies having something to do with it.

> I'm gonna say it again more specifically, I'm not talking about if the class is used or where it's used, I'm talking about this:
> We have a module which has one specific role: working with sms messages (sending them, receiving them, ...).

Who decided the specific role of the module? I thought our library was supposed to provide functions needed by killer-app in general, not just strictly regarding sending and receiving text messages.

> Does PreferenceManager have anything to do with that?

Yes, see the first part of this message.

> I'm saying it doesn't, can we both agree on this @giovannivelludo ?

No, I just explained why.

Cogno-IDU commented:
Alright, then I'll be even more specific, can we agree that that class has a **different functionality** than what the library is about (sending/receiving sms, ...)?

giovannivelludo commented:
When and by whom was it decided that the library used by killer-app had to provide only functionality relative to sending and receiving messages? If that's the case then I agree.

Cogno-IDU commented:
Uuuum, by the prof like a month ago? This is what we've been doing for a month ffs!

CremaLuca commented:
@giovannivelludo did you really just quote a dictionary reference to prove your point or am I dreaming?
And are you un-ironcally saying that the smslibrary should provide ability to read and write small chunks of data in memory?

giovannivelludo commented:
@Cogno-IDU 
> Uuuum, by the prof like a month ago? This is what we've been doing for a month ffs!

Ok, I just don't remember the part about the library having to include exclusively functionality related to sending and receiving SMS messages, in fact I remember doctor Li suggesting to make the library in a way that could be reused for sending those messages through other means, for example via Bluetooth; the package com.eis.communication includes classes that are not strictly related to sending and receiving SMS messages. If the library needs to contain exclusively functionality related to SMS messages, then that package should be moved to another module too.

> @giovannivelludo did you really just quote a dictionary reference to prove your point or am I dreaming?

I did.

> And are you un-ironcally saying that the smslibrary should provide ability to read and write small chunks of data in memory?

If smslibrary doesn't have to contain only functionality related to sending and receiving SMS messages, then yes. Otherwise no.

Cogno-IDU commented:
>  I just don't remember the part about the library having to include exclusively functionality related to sending and receiving SMS messages
Alright, then what about the functionality of sending and receiving data throw a communication channel (whichever that is)?
Does saving data in memory fall under that? I don't think so.

giovannivelludo commented:
@Cogno-IDU
It doesn't, but you just contradicted what you said above, which is that the only purpose of the library is sending and receiving SMSes. It's either that, or sending and receiving data through an arbitrary communication channel. It seems like you don't have a clear idea about what the library should do either.

EnricoCestaro commented:
@giovannivelludo 
> Ok, I just don't remember the part about the library having to include exclusively functionality related to sending and receiving SMS messages, in fact I remember doctor Li suggesting to make the library in a way that could be reused for sending those messages through other means, for example via Bluetooth

I don't think that the point of the tutor's suggestion was to make the library opened to other functionalities, I saw it as a tip about how to build the library itself, more like a guideline to explain how a class should be built.
The interfaces were in fact quite generic, they could be implemented by different classes, which could theoretically send or receive Bluetooth messages, SMS messages, or any other type of messages, with a dedicated class.
But the main point and purpose of the SMSLibrary are quite clear, sending and receiving SMS messages, all the killer apps were built upon this idea, and if there was a more appropriate time to discuss this, it certainly was a month ago. About this, you could simply answer me that we couldn't know if we needed other functionalities, and to that, I can answer that it clearly demonstrates what the library was meant to do since the beginning: the strictly necessary.

Cogno-IDU commented:
> It seems like you don't have a clear idea about what the library should do either.

Even if I don't have a clear idea of what the library does, I still have more idea that you by your own comments, which I quote:
> Ok, I just don't remember the part about the library having to include exclusively functionality related to sending and receiving SMS messages

and
> I thought our library was supposed to provide functions needed by killer-app in general, not just strictly regarding sending and receiving text messages.

So. Let's say the library is related to sending sms messages (which is reasonable, seeing that we've been working on this a whole month and counting). Now let's say that thanks to the tutor's help we extended this to sending data through a communication channel, which we kept specialyzing to just SMSMessages.
Now, does the PreferenceManager class respect this purpose? Well no, it's working on a different functionality, daving data to memory. And (since you've used a link to prove me wrong I'm gonna do the same) as written in the **official** documentation (which you seem to  really like based on this comment you wrote:

> I'm sorry if I sound rude, but I trust Android's documentation more than you

)
the **official** documentation states "A **module** is a collection of source files and build settings that allow you to divide your project into discrete units of functionality"
Does saving data make part of the functionality of sending messages (whicever they are)? No

giovannivelludo commented:
> Does saving data make part of the functionality of sending messages (whichever they are)? No

I had already agreed on that, I thought we were arguing about what the library should do now

Cogno-IDU commented:
> Does saving data make part of the functionality of sending messages (whichever they are)? No

> I had already agreed on that

Then why are you trying to force a class which has nothing to do with the functionality of the sms library inside the module of the library, while it should be in it's own separate module?

tonin-alessandra commented:
If I can give my opinion, I agree with people who said that preferences must be in a different place from sms library: they do different things and, in my opinion, preferences have nothing to do with sending or receiving sms, preferences and library have two different functions.

giovannivelludo commented:
@Cogno-IDU 
> Then why are you trying to force a class which has nothing to do with the functionality of the sms library inside the module of the library, while it should be in it's own separate module?

I'm not, I stopped after agreeing with you in [https://github.com/Cogno-IDU/smslibrary/pull/13#issuecomment-560566207](https://github.com/Cogno-IDU/smslibrary/pull/13#issuecomment-560566207)

giovannivelludo commented on smslibrary/src/main/AndroidManifest.xml:
```diff
@@ -5,6 +5,10 @@
+        <service
+            android:name=".background.SMSKeepAliveService"
+            android:enabled="true"
+            android:label="SMSKeepAliveService" />
```
I thought there was an agreement on removing the service since BroadcastReceivers declared in the manifest can be called by Android at any moment. Maybe it provides some other functionalities that I missed.

CremaLuca replied:
I really don't get why android refractor delete and build button don't tell me that I forgot to remove things around the project.
My fault indeed.

Cogno-IDU commented:
>Then why are you trying to force a class which has nothing to do with the functionality of the sms library inside the module of the library, while it should be in it's own separate module?

>I'm not, I stopped after agreeing with you in #13 (comment)

The comment says:

> It doesn't, but you just contradicted what you said above, which is that the only purpose of the library is sending and receiving SMSes. It's either that, or sending and receiving data through an arbitrary communication channel. It seems like you don't have a clear idea about what the library should do either.

I don't see any agreement, wrong link?

giovannivelludo commented:
@Cogno-IDU I agreed that saving data in memory doesn't fall under sending and receiving data through a communication channel

Cogno-IDU commented:
Which means putting that file somewhere else than in the smslibrery module itself. Alright, problem solved. Can this be merged now?

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/listeners/SMSReceivedServiceListener.java:
```diff
@@ -0,0 +1,47 @@
+ * @author Luca Crema, Alberto Ursino, Marco Tommasini, Alessandra Tonin
+ * @since 29/11/2019
+ */
+public abstract class SMSReceivedServiceListener extends IntentService {
```
Is this still needed after removing the service?

CremaLuca replied:
Sure it is, the app is dead and no activity can be woken up, so the app has to use a service that will run in the background without the need of the whole app to be loaded

giovannivelludo replied:
Ok, thanks

Cogno-IDU commented:
@CremaLuca There was a conflict that I tried to solve, but since I don't know what should be kept and what not it should have pushed the merge without the conflict in your branch, can you check if it all works?
This story never ends

giovannivelludo commented:
@Cogno-IDU this might have happened because pull requests made after this one were merged while this one is still open, and they probably modified lines that were modified in this pull request too. In the future we should try to review and merge pull requests in chronological order, to avoid making people go through the hassle of merging all the changes introduced after they opened their pull request. I realize that my requests for clarification are the main reason why this pull is being delayed, and I'm sorry for that.

CremaLuca commented:
@giovannivelludo, no problem, it's kinda normal to have pull request accepted in different orders, it's up to the author of the pull request (me) to keep up with the master
